src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -548,20 +548,34 @@ public async Task<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request,
                     ParseStatusLine(await ReadNextResponseHeaderLineAsync().ConfigureAwait(false), response);
                 }
 
-                // If we sent an Expect: 100-continue header, and didn't receive a 100-continue. Handle the final response accordingly.
-                // Note that the developer may have added an Expect: 100-continue header even if there is no Content.
+                // Parse the response headers.
+                while (true)
+                {
+                    ArraySegment<byte> line = await ReadNextResponseHeaderLineAsync(foldedHeadersAllowed: true).ConfigureAwait(false);
+                    if (IsLineEmpty(line))
+                    {
+                        break;
+                    }
+                    ParseHeaderNameValue(this, line, response);
+                }
+
                 if (allowExpect100ToContinue != null)
                 {
+                    // If we sent an Expect: 100-continue header, and didn't receive a 100-continue. Handle the final response accordingly.
+                    // Note that the developer may have added an Expect: 100-continue header even if there is no Content.
                     if ((int)response.StatusCode >= 300 &&
                         request.Content != null &&
-                        (request.Content.Headers.ContentLength == null || request.Content.Headers.ContentLength.GetValueOrDefault() > Expect100ErrorSendThreshold))
+                        (request.Content.Headers.ContentLength == null || request.Content.Headers.ContentLength.GetValueOrDefault() > Expect100ErrorSendThreshold) &&
+                        !((int)response.StatusCode == 401 && AuthenticationHelper.IsSessionAuthenticationChallenge(response)))
                     {
                         // For error final status codes, try to avoid sending the payload if its size is unknown or if it's known to be "big".
                         // If we already sent a header detailing the size of the payload, if we then don't send that payload, the server may wait
                         // for it and assume that the next request on the connection is actually this request's payload.  Thus we mark the connection
                         // to be closed.  However, we may have also lost a race condition with the Expect: 100-continue timeout, so if it turns out
                         // we've already started sending the payload (we weren't able to cancel it), then we don't need to force close the connection.
+                        // We also must not clone connection if we do NTLM or Negotiate authentication.
                         allowExpect100ToContinue.TrySetResult(false);
+
```

Comment:
GitHub won't let me comment on it, but the comment at L586 is now incomplete, right? e.g. it should be something like:
"For any success status codes, for errors when the request content length is known to be small, or for session-based authentication challenges, send the payload"
?